### DISCUSSION TITLE
What is PDL mean?

### DISCUSSION THREAD

**Discussion by MARD1NO** [NVIDIA/cutlass#1791] [Q&A] [answered]
I found there is a params `bool launch_with_pdl = false` in gemm universal adapter, what does it mean? It seems it is a Hopper feature

**Comment by thakkarV** [comment]
@alihassanijr

**Comment by alihassanijr** [comment, marked as answer]
PDL stands for [Programmatic Dependent Launch](https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#programmatic-dependent-launch-and-synchronization).

It is a feature new to Hopper through which one kernel that is currently being executed can allow dependent kernels to launch ahead of its completion and ramp up while the current kernel ramps down. Current kernel can issue `cudaTriggerProgrammaticLaunchCompletion()` when it is about to ramp down and finish execution, and the next kernel can gate its conflicting memory accesses with `cudaGridDependencySynchronize()`, which just blocks execution until the first kernel flushes its memory.

PDL requires both a kernel launch attribute and those device instructions built into the kernel, but only the former currently exists in CUTLASS.

**Reply by MARD1NO** [reply]
Thanks for quick reply! It seems it can overlap some kernel's launch and prepare works.

And I also have another question is: 

Assume A kernel has 2 Blocks, it just do a simple copy b[i] = a[i] like: 
```
__global__ void copy1(float* a, float* b, int32_t n) {
    #if (defined(__CUDA_ARCH__) && (__CUDA_ARCH__ >= 900))
        cudaGridDependencySynchronize();
    #endif

    for (int i = blockIdx.x * blockDim.x + threadIdx.x; i < n; i += blockDim.x * gridDim.x) {
        b[i] = a[i];
    }

    cudaTriggerProgrammaticLaunchCompletion();

}
```
then, B kernel need to consume the A kernel's output `b` pointer like(also take copy as an example): 
```
__global__ void copy2(float* b, float* c, int32_t n) {
    #if (defined(__CUDA_ARCH__) && (__CUDA_ARCH__ >= 900))
        cudaGridDependencySynchronize();
    #endif

    for (int i = blockIdx.x * blockDim.x + threadIdx.x; i < n; i += blockDim.x * gridDim.x) {
        c[i] = b[i];
    }

    cudaTriggerProgrammaticLaunchCompletion();

}
```
When copy1 kernel Block0 finish its copy work, can copy2 kernel Block0 start? Can PDL detect this pattern?

**Reply by MARD1NO** [reply]
And it seems `cudaTriggerProgrammaticLaunchCompletion` and `cudaGridDependencySynchronize()` (or its PTX)  is not used in cutlass code currently, it may not work ?

**Reply by alihassanijr** [reply]
Yes that is very much the idea; as long as there are SMs available for the next kernel to occupy, `cudaTriggerProgrammaticLaunchCompletion()` will unblock the next kernel in the stream and let it occupy SMs, and `cudaGridDependencySynchronize()` will have it (next kernel) wait until the previous kernel flushes its memory. This way copy1's results are visible to copy2 when it gets past `cudaGridDependencySynchronize()`.

**Reply by alihassanijr** [reply]
>And it seems cudaTriggerProgrammaticLaunchCompletion and cudaGridDependencySynchronize() (or its PTX) is not used in cutlass code currently, it may not work ?

Yes, I don't think any existing cutlass kernels have those instructions today. It is also very important to make sure all kernels launched with PDL and their next kernel have these instructions, especially `cudaGridDependencySynchronize()`, otherwise that can lead to incorrect results.

**Reply by MARD1NO** [reply]
So much thanks for your explanation!

**Reply by MARD1NO** [reply]
Hi, with the cutlass 3.6 published, I have noticed there exist an example to use PDL to prefetch matmul weight to overlap, but I still have some confusion. 

In cuda-c-programming documentation, it gives the example like that: 
![image](https://github.com/user-attachments/assets/74852143-5e63-4066-a986-696335ceeaec)

I just wonder if primary kernel has no work that can coincide with the secondary kernel like: 

(assume secondary kernel is hopper_pdl_gemm

```cpp
__global__ void primary_kernel() {
   // Initial work that should finish before starting secondary kernel

   // Trigger the secondary kernel
   cudaTriggerProgrammaticLaunchCompletion();

   // !!!! directly return
}
```
the secondary kernel will not be overlaped with primary kernel? or it still can overlap?

**Reply by MARD1NO** [reply]
> > And it seems cudaTriggerProgrammaticLaunchCompletion and cudaGridDependencySynchronize() (or its PTX) is not used in cutlass code currently, it may not work ?
> 
> Yes, I don't think any existing cutlass kernels have those instructions today. It is also very important to make sure all kernels launched with PDL and their next kernel have these instructions, especially `cudaGridDependencySynchronize()`, otherwise that can lead to incorrect results.

I test the pdl gemm exmaple kernel in my framework, when select a proper overlap ratio, I can observe the gemm kernel execution time is faster. But I do not observe any kernel overlap execution in nsys, does this meet the expectations?...

**Reply by alihassanijr** [reply]
>I just wonder if primary kernel has no work that can coincide with the secondary kernel

No, it's not that the primary kernel doesn't do any work. Primary kernel typically issues `cudaTriggerProgrammaticLaunchCompletion()` when it's about to finish its work. For example if you look at the [non-persistent warp specialized GEMM for hopper](https://github.com/NVIDIA/cutlass/blob/main/include/cutlass/gemm/kernel/sm90_gemm_tma_warpspecialized.hpp#L488-L491), you can see it's being called before the epilogue store, meaning the MMAs are done, and as it's about to write registers into gmem, it unblocks the next kernel from launching. However, that does not say anything about the next kernel occupying SMs.

>I test the pdl gemm exmaple kernel in my framework, when select a proper overlap ratio, I can observe the gemm kernel execution time is faster. But I do not observe any kernel overlap execution in nsys, does this meet the expectations?...

Did you build CUTLASS with `-DCUTLASS_ENABLE_GDC_FOR_SM90=1`?

**Reply by MARD1NO** [reply]
> > I just wonder if primary kernel has no work that can coincide with the secondary kernel
> 
> No, it's not that the primary kernel doesn't do any work. Primary kernel typically issues `cudaTriggerProgrammaticLaunchCompletion()` when it's about to finish its work. For example if you look at the [non-persistent warp specialized GEMM for hopper](https://github.com/NVIDIA/cutlass/blob/main/include/cutlass/gemm/kernel/sm90_gemm_tma_warpspecialized.hpp#L488-L491), you can see it's being called before the epilogue store, meaning the MMAs are done, and as it's about to write registers into gmem, it unblocks the next kernel from launching. However, that does not say anything about the next kernel occupying SMs.
> 
> > I test the pdl gemm exmaple kernel in my framework, when select a proper overlap ratio, I can observe the gemm kernel execution time is faster. But I do not observe any kernel overlap execution in nsys, does this meet the expectations?...
> 
> Did you build CUTLASS with `-DCUTLASS_ENABLE_GDC_FOR_SM90=1`?

Oh, I think I misinterpreted the PDL and example `63_hopper_gemm_with_weight_prefetch`. 

In [non-persistent warp specialized GEMM for hopper](https://github.com/NVIDIA/cutlass/blob/main/include/cutlass/gemm/kernel/sm90_gemm_tma_warpspecialized.hpp#L488-L491), you mean the mma warp finish the computation and can unblocks the next kernel, and the next kernel should add `wait_on_dependent_grids` when it need to consume previous kernel's output. there exist some opportunity to overlap the gemm epilogue and next kernel's perparation. 

In [63_hopper_gemm_with_weight_prefetch](https://github.com/NVIDIA/cutlass/tree/main/examples/63_hopper_gemm_with_weight_prefetch), I think it is used in like RMSNorm + Gemm case, when RMSNorm Kernel still runing, the gemm kernel  can prefetch the weight into L2 Cache, I construct this case, and ensure the previous kernel use pdl launch and invoke `cudaTriggerProgrammaticLaunchCompletion` before write normed result to global memory, but I do not observe the overlap execution. But its execution time is faster than the origin kernel(it means I set overlap ratio, prefetch ratio as -1.0)

And I'm sure I build this kernel with -DCUTLASS_ENABLE_GDC_FOR_SM90=1 and gemm.run with launch_pdl=True. 

Take a simple pdl demo as an example: 
```cpp
#include "cuda_runtime.h"


__global__ void copy1(float* __restrict__ a, float* __restrict__ b, 
                      int32_t n) {
    float tmp = 0.0f; 
                    
    for (int i = blockIdx.x * blockDim.x + threadIdx.x; i < n; i += blockDim.x * gridDim.x) {
        tmp += b[i];
    }

    cudaTriggerProgrammaticLaunchCompletion();

    for (int i = blockIdx.x * blockDim.x + threadIdx.x; i < n; i += blockDim.x * gridDim.x) {
        b[i] = a[i] + tmp;
    }

}

__global__ void copy2(const float* __restrict__ b, float* __restrict__ c, 
                      const float* __restrict__ d, int32_t n) {
    float result = 0.0f; 
                        
    for (int i = blockIdx.x * blockDim.x + threadIdx.x; i < n; i += blockDim.x * gridDim.x) {
        result += d[i];
    }

    cudaGridDependencySynchronize();

    for (int i = blockIdx.x * blockDim.x + threadIdx.x; i < n; i += blockDim.x * gridDim.x) {
        c[i] = b[i] + result + 2.0f;
    }

}


int main() {
    // const int n = 4096 * 4096; 
    const int n = 64 * 4096; 

    float *a, *b, *c, *d;
    cudaMalloc(&a, n * sizeof(float));
    cudaMalloc(&b, n * sizeof(float));
    cudaMalloc(&c, n * sizeof(float));
    cudaMalloc(&d, n * sizeof(float));

    const int32_t grid = 16; 
    const int32_t blockdim = 256; 


    cudaStream_t stream; 
    cudaStreamCreate(&stream);

    for (int i = 0; i < 10; i++) {
        cudaLaunchConfig_t kernelConfig2 = {0};                                                         
        kernelConfig2.gridDim = grid;   
        kernelConfig2.blockDim = blockdim;                                                                
        kernelConfig2.dynamicSmemBytes = 0;                                                   
        kernelConfig2.stream = stream;                                                                
        cudaLaunchAttribute attribute2[1];                                                            
        attribute2[0].id = cudaLaunchAttributeProgrammaticStreamSerialization;                        
        attribute2[0].val.programmaticStreamSerializationAllowed = 1;                                 
        kernelConfig2.attrs = attribute2;                                                              
        kernelConfig2.numAttrs = 1;  

        // PDL Launch
        // copy1<<<grid, blockdim, 0, stream>>>(a, b, n);  
        // cudaLaunchKernelEx(&kernelConfig2, &copy2, b, c, d, n);  

        // No PDL Launch
        copy1<<<grid, blockdim, 0, stream>>>(a, b, n);  
        copy2<<<grid, blockdim, 0, stream>>>(b, c, d, n);  
    }
    
    cudaDeviceSynchronize(); 

    cudaFree(a); 
    cudaFree(b); 
    cudaFree(c); 
    cudaFree(d);

    cudaStreamDestroy(stream);
}

```
When the elem_cnt is 4096 * 4096, I can observe kernel execution overlap. But in 64 * 4096 cases, I do not observe overlap, But the end2end time is faster with pdl launch: 

PDL launch nsys: 40us
![image](https://github.com/user-attachments/assets/5004ae08-cc95-4fb9-a912-6f5a86d31be5)

origin launch: 50us
![image](https://github.com/user-attachments/assets/9eebf03e-e358-4fe3-b1c6-c348d147d8bc)

Is this expected or I make some mistake?...

here is my nsys profile command:
```
nvcc -arch=sm_90a -O3 test3.cu -o test

export CUDA_VISIBLE_DEVICES=0
nsys profile --force-overwrite=true --cuda-graph-trace=node --stats=true \
    -o test_overlap ./test

```

**Reply by alihassanijr** [reply]
>In non-persistent warp specialized GEMM for hopper, you mean the mma warp finish the computation and can unblocks the next kernel, and the next kernel should add wait_on_dependent_grids when it need to consume previous kernel's output. there exist some opportunity to overlap the gemm epilogue and next kernel's perparation.

Yes exactly.

>In 63_hopper_gemm_with_weight_prefetch, I think it is used in like RMSNorm + Gemm case, when RMSNorm Kernel still runing, the gemm kernel  can prefetch the weight into L2 Cache, I construct this case, and ensure the previous kernel use pdl launch and invoke cudaTriggerProgrammaticLaunchCompletion before write normed result to global memory, but I do not observe the overlap execution. But its execution time is faster than the origin kernel(it means I set overlap ratio, prefetch ratio as -1.0)

Yes that is the idea. Regarding runtime, I think there's a bug in the example. L2 cache lines should be cleared, otherwise as long as you prefetch once, the operands could very well become L2 resident and throw off the entire benchmark.

How much of a difference in effective BW / runtime are you observing?

>Is this expected or I make some mistake?...

Sorry I'm confused, so if you are seeing overlap then what's the issue?

Keep in mind that the overlap is not in any way guaranteed, the same way that concurrency of launched kernels is not guaranteed.

**Reply by MARD1NO** [reply]
Thanks for your explanation, I think it is clear. :D

**Reply by chenzhanyiczy2** [reply]
> > > I just wonder if primary kernel has no work that can coincide with the secondary kernel
> > 
> > 
> > No, it's not that the primary kernel doesn't do any work. Primary kernel typically issues `cudaTriggerProgrammaticLaunchCompletion()` when it's about to finish its work. For example if you look at the [non-persistent warp specialized GEMM for hopper](https://github.com/NVIDIA/cutlass/blob/main/include/cutlass/gemm/kernel/sm90_gemm_tma_warpspecialized.hpp#L488-L491), you can see it's being called before the epilogue store, meaning the MMAs are done, and as it's about to write registers into gmem, it unblocks the next kernel from launching. However, that does not say anything about the next kernel occupying SMs.
> > > I test the pdl gemm exmaple kernel in my framework, when select a proper overlap ratio, I can observe the gemm kernel execution time is faster. But I do not observe any kernel overlap execution in nsys, does this meet the expectations?...
> > 
> > 
> > Did you build CUTLASS with `-DCUTLASS_ENABLE_GDC_FOR_SM90=1`?
> 
> Oh, I think I misinterpreted the PDL and example `63_hopper_gemm_with_weight_prefetch`.
> 
> In [non-persistent warp specialized GEMM for hopper](https://github.com/NVIDIA/cutlass/blob/main/include/cutlass/gemm/kernel/sm90_gemm_tma_warpspecialized.hpp#L488-L491), you mean the mma warp finish the computation and can unblocks the next kernel, and the next kernel should add `wait_on_dependent_grids` when it need to consume previous kernel's output. there exist some opportunity to overlap the gemm epilogue and next kernel's perparation.
> 
> In [63_hopper_gemm_with_weight_prefetch](https://github.com/NVIDIA/cutlass/tree/main/examples/63_hopper_gemm_with_weight_prefetch), I think it is used in like RMSNorm + Gemm case, when RMSNorm Kernel still runing, the gemm kernel  can prefetch the weight into L2 Cache, I construct this case, and ensure the previous kernel use pdl launch and invoke `cudaTriggerProgrammaticLaunchCompletion` before write normed result to global memory, but I do not observe the overlap execution. But its execution time is faster than the origin kernel(it means I set overlap ratio, prefetch ratio as -1.0)
> 
> And I'm sure I build this kernel with -DCUTLASS_ENABLE_GDC_FOR_SM90=1 and gemm.run with launch_pdl=True.
> 
> Take a simple pdl demo as an example:
> 
> ```c++
> #include "cuda_runtime.h"
> 
> 
> __global__ void copy1(float* __restrict__ a, float* __restrict__ b, 
>                       int32_t n) {
>     float tmp = 0.0f; 
>                     
>     for (int i = blockIdx.x * blockDim.x + threadIdx.x; i < n; i += blockDim.x * gridDim.x) {
>         tmp += b[i];
>     }
> 
>     cudaTriggerProgrammaticLaunchCompletion();
> 
>     for (int i = blockIdx.x * blockDim.x + threadIdx.x; i < n; i += blockDim.x * gridDim.x) {
>         b[i] = a[i] + tmp;
>     }
> 
> }
> 
> __global__ void copy2(const float* __restrict__ b, float* __restrict__ c, 
>                       const float* __restrict__ d, int32_t n) {
>     float result = 0.0f; 
>                         
>     for (int i = blockIdx.x * blockDim.x + threadIdx.x; i < n; i += blockDim.x * gridDim.x) {
>         result += d[i];
>     }
> 
>     cudaGridDependencySynchronize();
> 
>     for (int i = blockIdx.x * blockDim.x + threadIdx.x; i < n; i += blockDim.x * gridDim.x) {
>         c[i] = b[i] + result + 2.0f;
>     }
> 
> }
> 
> 
> int main() {
>     // const int n = 4096 * 4096; 
>     const int n = 64 * 4096; 
> 
>     float *a, *b, *c, *d;
>     cudaMalloc(&a, n * sizeof(float));
>     cudaMalloc(&b, n * sizeof(float));
>     cudaMalloc(&c, n * sizeof(float));
>     cudaMalloc(&d, n * sizeof(float));
> 
>     const int32_t grid = 16; 
>     const int32_t blockdim = 256; 
> 
> 
>     cudaStream_t stream; 
>     cudaStreamCreate(&stream);
> 
>     for (int i = 0; i < 10; i++) {
>         cudaLaunchConfig_t kernelConfig2 = {0};                                                         
>         kernelConfig2.gridDim = grid;   
>         kernelConfig2.blockDim = blockdim;                                                                
>         kernelConfig2.dynamicSmemBytes = 0;                                                   
>         kernelConfig2.stream = stream;                                                                
>         cudaLaunchAttribute attribute2[1];                                                            
>         attribute2[0].id = cudaLaunchAttributeProgrammaticStreamSerialization;                        
>         attribute2[0].val.programmaticStreamSerializationAllowed = 1;                                 
>         kernelConfig2.attrs = attribute2;                                                              
>         kernelConfig2.numAttrs = 1;  
> 
>         // PDL Launch
>         // copy1<<<grid, blockdim, 0, stream>>>(a, b, n);  
>         // cudaLaunchKernelEx(&kernelConfig2, &copy2, b, c, d, n);  
> 
>         // No PDL Launch
>         copy1<<<grid, blockdim, 0, stream>>>(a, b, n);  
>         copy2<<<grid, blockdim, 0, stream>>>(b, c, d, n);  
>     }
>     
>     cudaDeviceSynchronize(); 
> 
>     cudaFree(a); 
>     cudaFree(b); 
>     cudaFree(c); 
>     cudaFree(d);
> 
>     cudaStreamDestroy(stream);
> }
> ```
> 
> When the elem_cnt is 4096 * 4096, I can observe kernel execution overlap. But in 64 * 4096 cases, I do not observe overlap, But the end2end time is faster with pdl launch:
> 
> PDL launch nsys: 40us ![image](https://private-user-images.githubusercontent.com/42901638/376104295-5004ae08-cc95-4fb9-a912-6f5a86d31be5.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzYzMTU3NjksIm5iZiI6MTczNjMxNTQ2OSwicGF0aCI6Ii80MjkwMTYzOC8zNzYxMDQyOTUtNTAwNGFlMDgtY2M5NS00ZmI5LWE5MTItNmY1YTg2ZDMxYmU1LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAxMDglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMTA4VDA1NTEwOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTE3MGFiZDRmMzNjNzczYzMwNTk5YjdlNDY4YmQxNTY4NTEyN2RkNzllYWIxMmM0YzQ5ZTRhYWM5NzhhZDkyNDQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.VIiueJgdmp2ww7MdPpVfCMYdtrHCH1kNEoV9urPtHMk)
> 
> origin launch: 50us ![image](https://private-user-images.githubusercontent.com/42901638/376104362-9eebf03e-e358-4fe3-b1c6-c348d147d8bc.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzYzMTU3NjksIm5iZiI6MTczNjMxNTQ2OSwicGF0aCI6Ii80MjkwMTYzOC8zNzYxMDQzNjItOWVlYmYwM2UtZTM1OC00ZmUzLWIxYzYtYzM0OGQxNDdkOGJjLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAxMDglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMTA4VDA1NTEwOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTJkNjYzOGVjY2E1ZGZhYzQyMjEzMWM4NzdkN2ZmYWFkNzRkNTBiOTlhZWRjZTUzZWUyZjBlZWRiNDM1MzBiZTkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.d2tzxodYNtvC7TEC89ZUD0_Yjj7o0AxSYdFVABtAcaI)
> 
> Is this expected or I make some mistake?...
> 
> here is my nsys profile command:
> 
> ```
> nvcc -arch=sm_90a -O3 test3.cu -o test
> 
> export CUDA_VISIBLE_DEVICES=0
> nsys profile --force-overwrite=true --cuda-graph-trace=node --stats=true \
>     -o test_overlap ./test
> ```

I also used the 63 example, but did not find any acceleration effect. How did you test the rmsnorm+63 example?  thanks.
